File: eel-core/src/main/kotlin/io/eels/component/jdbc/ResultsetPart.kt
package io.eels.component.jdbc

import io.eels.util.Logging
import io.eels.Row
import io.eels.schema.Schema
import io.eels.component.Part
import rx.Observable
import java.sql.Connection
import java.sql.ResultSet
import java.sql.Statement

/**
 * A Part for a Resultset. Will publish all rows from the resultset and then close the resultset.
 */
class ResultsetPart(val rs: ResultSet,
                    val stmt: Statement,
                    val conn: Connection,
                    val schema: Schema) : Part, Logging {

  override fun data(): Observable<Row> {
    return Observable.create<Row> {
      try {
        it.onStart()
        while (rs.next()) {
          val values = schema.columnNames().map { rs.getObject(it) }
          val row = Row(schema, values)
          it.onNext(row)
        }
        it.onCompleted()
      } catch(t: Throwable) {
        it.onError(t)
      } finally {
        conn.close()
      }
    }
  }
}

File: eel-core/src/main/scala/io/eels/component/jdbc/ResultsetPart.scala
package io.eels.component.jdbc

import java.sql.{Connection, ResultSet, Statement}

import com.sksamuel.exts.Logging
import io.eels.schema.Schema
import io.eels.{Part, Row, RowListener}
import rx.lang.scala.Observable

/**
 * A Part for a Resultset. Will publish all rows from the resultset and then close the resultset.
 */
class ResultsetPart(val rs: ResultSet,
                    val stmt: Statement,
                    val conn: Connection,
                    val schema: Schema,
                    val listener: RowListener) extends Part with Logging {

  override def data(): Observable[Row] = {
    Observable.apply { subscriber =>
      try {
        subscriber.onStart()
        while (rs.next()) {
          val values = schema.fieldNames().map(name => rs.getObject(name))
          val row = Row(schema, values)
          subscriber.onNext(row)
          listener.onRow(row)
        }
        subscriber.onCompleted()
      } catch {
        case t: Throwable =>
          subscriber.onError(t)
      } finally {
        conn.close()
      }
    }
  }
}
